File: pandascript-vs-cdp/scripts/playwright/news.js
import { chromium } from "playwright-core";

const endpoint = process.env.BROWSER_WS ?? "ws://127.0.0.1:9222";
const browser = await chromium.connectOverCDP(endpoint);
const context = await browser.newContext();
const page = await context.newPage();

await page.goto("https://apnews.com/hub/world-news");

const links = await page.$$eval("a[href*='/article/']", (anchors) =>
  anchors.slice(0, 10).map((a) => a.href),
);
const urls = [...new Set(links)].slice(0, 3);

const articles = [];
for (const url of urls) {
  await page.goto(url);
  await page.waitForSelector(".RichTextStoryBody p");
  const headline = await page.$eval("h1", (h) => h.textContent.trim());
  const paragraphs = await page.$$eval(".RichTextStoryBody p", (ps) =>
    ps.slice(0, 3).map((p) => p.textContent.trim()),
  );
  articles.push({ url, headline, paragraphs });
}

console.log(JSON.stringify(articles));

await page.close();
await context.close();
await browser.close();

File: pandascript-vs-cdp/scripts/playwright-py/news.py
import json
import os

from playwright.sync_api import sync_playwright

endpoint = os.environ.get("BROWSER_WS", "ws://127.0.0.1:9222")

with sync_playwright() as p:
    browser = p.chromium.connect_over_cdp(endpoint)
    context = browser.new_context()
    page = context.new_page()

    page.goto("https://apnews.com/hub/world-news")

    links = page.eval_on_selector_all(
        "a[href*='/article/']",
        "(anchors) => anchors.slice(0, 10).map((a) => a.href)",
    )
    urls = list(dict.fromkeys(links))[:3]

    articles = []
    for url in urls:
        page.goto(url)
        page.wait_for_selector(".RichTextStoryBody p")
        headline = page.eval_on_selector("h1", "(h) => h.textContent.trim()")
        paragraphs = page.eval_on_selector_all(
            ".RichTextStoryBody p",
            "(ps) => ps.slice(0, 3).map((p) => p.textContent.trim())",
        )
        articles.append({"url": url, "headline": headline, "paragraphs": paragraphs})

    print(json.dumps(articles))

    page.close()
    context.close()
    browser.close()
